reference/test/matrix/csr_kernels.cpp

```diff
@@ -1284,6 +1310,439 @@ TYPED_TEST(Csr, NonSquareMtxIsTransposable)
 }
 
 
+template <typename ValueType, typename IndexType>
+std::unique_ptr<gko::matrix::Csr<ValueType, IndexType>> csr_from_permutation(
+    gko::matrix::Permutation<IndexType>* perm, bool invert)
+{
+    gko::matrix_data<double, IndexType> double_data;
+    if (invert) {
+        perm->invert()->write(double_data);
+    } else {
+        perm->write(double_data);
+    }
+    gko::matrix_data<ValueType, IndexType> data;
+    data.size = double_data.size;
+    for (auto entry : double_data.nonzeros) {
+        data.nonzeros.emplace_back(entry.row, entry.column, 1.0);
+    }
+    auto mtx =
+        gko::matrix::Csr<ValueType, IndexType>::create(perm->get_executor());
+    mtx->read(data);
+    return mtx;
+}
+
+
+template <typename ValueType, typename IndexType>
+std::unique_ptr<gko::matrix::Csr<ValueType, IndexType>> csr_from_permutation(
+    gko::matrix::ScaledPermutation<ValueType, IndexType>* perm, bool invert)
+{
+    gko::matrix_data<ValueType, IndexType> data;
+    if (invert) {
+        perm->invert()->write(data);
+    } else {
+        perm->write(data);
+    }
+    auto mtx =
+        gko::matrix::Csr<ValueType, IndexType>::create(perm->get_executor());
+    mtx->read(data);
+    return mtx;
+}
+
+
+template <typename ValueType, typename IndexType, typename Permutation>
+std::unique_ptr<gko::matrix::Csr<ValueType, IndexType>> ref_permute(
+    gko::matrix::Csr<ValueType, IndexType>* input, Permutation* permutation,
+    gko::matrix::permute_mode mode)
+{
+    using gko::matrix::permute_mode;
+    using Csr = gko::matrix::Csr<ValueType, IndexType>;
+    auto result = input->clone();
+    auto permutation_csr = csr_from_permutation<ValueType>(
+        permutation, (mode & permute_mode::inverse) == permute_mode::inverse);
+    if ((mode & permute_mode::rows) == permute_mode::rows) {
+        // compute P * A
+        permutation_csr->apply(input, result);
```

Comment:
should we make these tests rely on SpGEMM?

Reply:
I think so, that's the easiest way to check that the algorithms are correct, since SpGEMM in reference is fairly simple.